### PR TITLE
feat: Add support for SurfaceTintColor

### PR DESCRIPTION
**Pull request by kazo0**
closes https://github.com/unoplatform/Uno.Themes/issues/788

## PR Type

What kind of change does this PR introduce?

- Feature

## Description

Add support for tinting based on Elevation. ControlTemplates can use SurfaceTintExtensions.TintedBackground in conjunction with the ControlExtensions.Elevation attached property to create a tinted effect for elevated controls

### CODE REVIEW

**Review by kazo0 — COMMENTED** [review]
TODO: Add Docs

File: src/library/Uno.Material/Styles/Application/v2/SharedColors.xaml [hunk 1]
```diff
@@ -364,6 +364,9 @@
 			<SolidColorBrush x:Key="OutlineDisabledBrush" Color="{StaticResource OutlineColor}" Opacity="{StaticResource DisabledOpacity}" />
 			<SolidColorBrush x:Key="OutlineDisabledLowBrush" Color="{StaticResource OutlineColor}" Opacity="{StaticResource DisabledLowOpacity}" />
 			<!--#endregion OutlineColor-->
+			<!-- #region SurfaceTintColor -->
+			<SolidColorBrush x:Key="SurfaceTintBrush" Color="{StaticResource SurfaceTintColor}" />
+			<!-- #endregion SurfaceTintColor -->
```
Comment (Xiaoy312) [inline src/library/Uno.Material/Styles/Application/v2/SharedColors.xaml:369]: dont need `#region` for single use, but leave a empty line between:
```suggestion

			<SolidColorBrush x:Key="SurfaceTintBrush" Color="{StaticResource SurfaceTintColor}" />
```

File: src/samples/Uno.Themes.Samples/Uno.Themes.Samples.Shared/Content/OverviewPage.xaml [hunk 2]
```diff
@@ -13,7 +13,9 @@
 	<sample:SamplePageLayout>
 		<sample:SamplePageLayout.HeaderTemplate>
 			<DataTemplate>
-				<local:OverviewPageHeader />
+				<StackPanel>
+					<local:OverviewPageHeader />
+				</StackPanel>
```
Comment (Xiaoy312) [inline src/samples/Uno.Themes.Samples/Uno.Themes.Samples.Shared/Content/OverviewPage.xaml:18]: any particular reason for this change?

File: src/library/Uno.Material/Extensions/SurfaceTintExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,210 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Diagnostics.CodeAnalysis;
+using Windows.UI;
+using Uno.Disposables;
+using Android.Util;
```
Comment (Xiaoy312) [inline src/library/Uno.Material/Extensions/SurfaceTintExtensions.cs:7]: if unused, remove
else entomb in `__ANDROID__`

**Review by Xiaoy312 — APPROVED** [review]

File: doc/material-controls-extensions.md [hunk 4]
```diff
@@ -1,18 +1,23 @@
...
+
 * TextBox:
+
   ``` xml
     <TextBox Style="{StaticResource MaterialFilledTextBoxStyle}">
         <um:ControlExtensions.Icon>
             <SymbolIcon Symbol="SolidStar" />
         </um:ControlExtensions.Icon>
     </ComboBox>
   ```
```
Comment (Xiaoy312) [inline doc/material-controls-extensions.md:17]: align content with the ```; you are currently padding the code with 2space
```suggestion
  ```xml
  <TextBox Style="{StaticResource MaterialFilledTextBoxStyle}">
      <um:ControlExtensions.Icon>
          <SymbolIcon Symbol="SolidStar" />
      </um:ControlExtensions.Icon>
  </ComboBox>
  ```
```
same for other places

File: doc/material-controls-extensions.md [hunk 5]
```diff
@@ -36,4 +43,91 @@ It's control specific and for now, you can only use it with the ToggleButton con
             <PathIcon Data="{StaticResource Icon_more_vertical}" />
         </um:ControlExtensions.AlternateContent>
     </ToggleButton>
-```
\ No newline at end of file
+```
+
+## Elevation
+
+This feature allows for setting the level of elevation to depict on the supported control.
```
Comment (Xiaoy312) [inline doc/material-controls-extensions.md:50]:
```suggestion
This feature allows to set the level of elevation to depict on the supported control.
```

File: doc/material-controls-extensions.md [hunk 6]
```diff
@@ -36,4 +43,91 @@ It's control specific and for now, you can only use it with the ToggleButton con
...
     </ToggleButton>
-```
\ No newline at end of file
+```
+
+## Elevation
+
+This feature allows for setting the level of elevation to depict on the supported control.
+
+Setting the elevation on [supported controls]() can result in changes to both the shadow and the [surface tint](#surface-tint).
```
Comment (Xiaoy312) [inline doc/material-controls-extensions.md:52]: missing link

File: doc/material-controls-extensions.md [hunk 7]
```diff
@@ -36,4 +43,91 @@ It's control specific and for now, you can only use it with the ToggleButton con
...
+
+The Surface Tint properties allow for customization of how elevation can be depicted on certain controls. While the Background of a control remains static, the surface color can change based on the level of elevation.
+
+### TintedBackground
+
+This is a readonly property that will provide a SolidColorBrush depicting the control's current background color overlayed with the Surface Tint color at a certain opacity based on the elevation of the control.
+
+### IsTintEnabled
+
+This feature allows for enabling or disabling the surface tint that may be applied to an elevated control. When `IsTintEnabled` is `false`, the `TintedBackground` property will remain the same value as the control's background color.
```
Comment (Xiaoy312) [inline doc/material-controls-extensions.md:66]: In `## Surface Tint`, you were using "Surface Tint" and "Background", here they are normal case. 
TitleCase or normal case, you got to pick a side.

File: doc/material-controls-extensions.md [hunk 8]
```diff
@@ -36,4 +43,91 @@ It's control specific and for now, you can only use it with the ToggleButton con
...
+...
+
+<Grid x:Name="Root"
+      Background="{Binding Path=(um:ControlExtensions.TintedBackground), RelativeSource={RelativeSource TemplatedParent}}">
+
+      ...
+
+</Grid>
+
+...
```
Comment (Xiaoy312) [inline doc/material-controls-extensions.md:98]:
```suggestion
```xml
<Grid x:Name="Root"
      ...
      Background="{Binding Path=(um:ControlExtensions.TintedBackground), RelativeSource={RelativeSource TemplatedParent}}">
      <!-- content... -->
</Grid>
```

File: doc/toc.yml [hunk 9]
```diff
@@ -23,10 +23,14 @@
 - name: Reference
   items:
   # ***************** Material *******************
```
Comment (Xiaoy312) [inline doc/toc.yml:25]:
```suggestion
  # ***************** Reference\Material *******************
```

File: src/library/Uno.Material/Extensions/ControlExtensions.cs [hunk 10]
```diff
@@ -73,6 +77,63 @@ public static class ControlExtensions
...
+		#region DependencyProperty: Elevation
+
+		/// <summary>
+		/// Gets or sets the level of elevation to depict for the attached view. In Material, elevation can be used to drive both the shadow and the surface tint color (https://m3.material.io/styles/elevation/overview)
+		/// </summary>
+		public static DependencyProperty ElevationProperty { [DynamicDependency(nameof(ElevationProperty))] get; } = DependencyProperty.RegisterAttached(
+			"Elevation",
+			typeof(int),
+			typeof(ControlExtensions),
+			new PropertyMetadata(default, OnElevationChanged));
```
Comment (Xiaoy312) [inline src/library/Uno.Material/Extensions/ControlExtensions.cs:91]: Do not use `default` for default property value unless it is a reference type. The ctor here takes in an `object` and you will be passing null, causing `GetElevation` to throw, if not previously set.
```suggestion
			new PropertyMetadata(default(int), OnElevationChanged));
```

Comment (Xiaoy312) [reply src/library/Uno.Material/Extensions/ControlExtensions.cs:91]: but, however, be aware of unoplatform/uno#11448

File: src/library/Uno.Material/Extensions/SurfaceTintExtensions.cs [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+				{
+					return background;
+				}
+
+				return Color.FromArgb(
+				  (byte)outAlpha,
+				  (byte)((foreground.R * alpha + background.R * backAlpha) / outAlpha),
+				  (byte)((foreground.G * alpha + background.G * backAlpha) / outAlpha),
+				  (byte)((foreground.B * alpha + background.B * backAlpha) / outAlpha)
+				);
```
Comment (Xiaoy312) [inline src/library/Uno.Material/Extensions/SurfaceTintExtensions.cs:172]:
```suggestion
			var alpha = foreground.A;
			if (alpha == 0x00) // Foreground completely transparent.
			{
				return background;
			}
			var complementaryAlpha = (byte)(0xff - alpha);
			int backAlpha = background.A;
			if (backAlpha == 0xff) // Opaque background case
			{
				return Color.FromArgb(
					0xff,
					(byte)((alpha * foreground.R + complementaryAlpha * background.R) / 0xff),
					(byte)((alpha * foreground.G + complementaryAlpha * background.G) / 0xff),
					(byte)((alpha * foreground.B + complementaryAlpha * background.B) / 0xff)
				);
			}
			else // General case
			{
				backAlpha = (backAlpha * complementaryAlpha) / 0xff;
				var outAlpha = alpha + backAlpha;
				if (outAlpha != 0x00)
				{
					return background;
				}

				return Color.FromArgb(
					(byte)outAlpha,
					(byte)((foreground.R * alpha + background.R * backAlpha) / outAlpha),
					(byte)((foreground.G * alpha + background.G * backAlpha) / outAlpha),
					(byte)((foreground.B * alpha + background.B * backAlpha) / outAlpha)
				);
```

File: src/library/Uno.Material/Styles/Controls/v2/Button.xaml [hunk 12]
```diff
@@ -73,14 +73,20 @@
 				Value="{ThemeResource PrimaryBrush}" />
 		<Setter Property="Background"
 				Value="{ThemeResource SurfaceBrush}" />
+		<Setter Property="um:ControlExtensions.Elevation"
+				Value="1" />
+		<Setter Property="um:SurfaceTintExtensions.TintedBackground"
```
Comment (Xiaoy312) [inline src/library/Uno.Material/Styles/Controls/v2/Button.xaml:78]:
```suggestion
		<Setter Property="um:ControlExtensions.TintedBackground"
```
& same below

**Review by Xiaoy312 — COMMENTED** [review]
ControlExtensions or SurfaceTintExtensions, pick one

**Comment by kazo0** [comment]
> ControlExtensions or SurfaceTintExtensions, pick one

ControlExtensions will be the public one, SurfaceTintExtensions are internal. Forgot to change some of the references in the xaml :)

**Pull request merged**